_posts/timeouts-power.md

```diff
@@ -0,0 +1,74 @@
+
+## The diagnose
+
+The [postmortem](https://en.wikipedia.org/wiki/Postmortem_documentation) discussion focused on why the service initiating the order checkout process failed. Once that was settled, the attention moved to the payment service, and coupling entered the scene. 
+
+The team started wondering why Payments did not cancel the card authorization. The authorization and charging process was configured to react to other service failures and cancel any previously issued authorization requests.
+
+It took a couple of seconds to realize the colossal mistake at play. All those requests to other teams to publish non-business-related events to signal failures and errors caused an intricated spaghetti-style web of cross-service relationships.
+
+The Payment service depends on the Order checkout to revert the card authorization. They were coupled instead of being autonomous. [It was one of those instances where the team's interpretation of the word autonomous led in the wrong direction](https://milestone.topics.it/2022/09/05/autonomy.html).
+
+Customers failed to proceed with the checkout process because the bug was causing an outage in the Order checkout service. The service couldn't publish any message, including those meant to signal the process failure. From the Payments service perspective, no news was good news, and it kept the card authorization in place.
+
+## The cure
+
+The group in charge started analyzing what they could have done differently. How can they change services to prevent something like this from happening in the future? In one word: how to untangle the knot?
+
+One option is to leverage the power of timeouts, also known as delayed deliveries. I like to describe them as a neat way to react to things _not_ happening. That's precisely the case at play.
+
+A timeout is a message sent in the future. The sender sets the desired delivery date and time, and the infrastructure delivers the message when requested.
+
+If the Order checkout process fails, the Payments service wants to retract the card authorization. That's the way the business describes the thing.
+
+We can look at it from the perspective of what's _not_ happening. The Payments service wants to retract the cart authorization if _it doesn't receive any signal from the Orders checkout within a particular time from the approval_.
+
+For example, if on average (and be aware of the [risk of using averages](https://towardsdatascience.com/why-averages-are-often-wrong-1ff08e409a5b)) the Order checkout process publishes an even to signal its completion in 10 minutes, we could say that if within a day we heard nothing it's fair to assume that something went so wrong that it's better to cancel the card authorization rather than insiting on holding the money.
```

Comment:
```suggestion
For example, if on average (and be aware of the [risk of using averages](https://towardsdatascience.com/why-averages-are-often-wrong-1ff08e409a5b)) the Order checkout process publishes an event to signal its completion in 10 minutes, we could say that if within a day we heard nothing it's fair to assume that something went wrong and it's better to cancel the card authorization rather than holding the money.
```